### PR TITLE
issue: 843662 on bond interface check preconditions for offload on al…

### PR DESCRIPTION
**Pull request by dinal**
…l slaves

during device mapping on startup we should check preconditions on all active slaves

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/329/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/330/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/311/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/312/ for details (Mellanox internal link).

File: src/vma/dev/net_device_table_mgr.h [hunk 1]
```diff
@@ -135,6 +135,7 @@ class net_device_table_mgr : public cache_table_mgr<ip_address, net_device_val*>
 	bool 			verify_ipoib_mode(struct ifaddrs* ifa);
 	bool			verify_mlx4_ib_device(const char* ifname);
 	bool 			verify_eth_qp_creation(const char* ifname);
+	bool 			check_preconditions(const char* interface_name, struct ifaddrs * ifa);
```
Comment (OphirMunk) [inline src/vma/dev/net_device_table_mgr.h]: All methods are verify_XXX.
Therefore need to rename check_preconditions() to verify_preconditions()

Comment (OphirMunk) [inline src/vma/dev/net_device_table_mgr.h]: More accurately: rename check_preconditions() to verify_ipoib_or_eth_qp_creation()

File: src/vma/dev/net_device_table_mgr.cpp [hunk 2]
```diff
@@ -243,16 +243,38 @@ int net_device_table_mgr::map_net_devices()
...
+				char* p = strchr(slave_name, '\n');
+				if (p) *p = '\0'; // Remove the tailing 'new line" char
+				if (!check_preconditions(slave_name, ifa)) {
+					valid = false;
+					vlog_printf(VLOG_WARNING,"******************************************************\n");
+					vlog_printf(VLOG_WARNING,"* Bond %s will not be offloaded.\n", ifa->ifa_name);
+					vlog_printf(VLOG_WARNING,"* Slave %s did not meet requirements.\n", slave_name);
+					vlog_printf(VLOG_WARNING,"* Check warning messages for more information.\n");
+					vlog_printf(VLOG_WARNING,"******************************************************\n");
+					break;
```
Comment (OphirMunk) [inline src/vma/dev/net_device_table_mgr.cpp]: This function map_net_devices() calls other functions which print WARNINGs. Please keep this hierarchy by calling a wrapper function for the printing

File: src/vma/dev/net_device_table_mgr.cpp [hunk 3]
```diff
@@ -243,16 +243,38 @@ int net_device_table_mgr::map_net_devices()
...
 
-		int iftype = get_iftype_from_ifname(ifa->ifa_name);
 		bool valid = false;
-		if (iftype == ARPHRD_INFINIBAND) {
-			if (verify_enable_ipoib(ifa->ifa_name) && verify_mlx4_ib_device(ifa->ifa_name) && verify_ipoib_mode(ifa)) {
-				valid = true;
+		if (ifa->ifa_flags & IFF_MASTER) {
+			// this is a bond interface, find the slave
+			char active_slave_name[IFNAMSIZ * MAX_SLAVES] = {0};
+			if (!get_bond_active_slave_name(ifa->ifa_name, active_slave_name, sizeof active_slave_name)) {
```
Comment (OphirMunk) [inline src/vma/dev/net_device_table_mgr.cpp]: Please explain why you only verify the active slaves and skip checking all the others. Shouldn't you verify all slaves?
Is it possible that one slave is IB and the other one is ETH?

File: src/vma/dev/net_device_table_mgr.cpp [hunk 4]
```diff
@@ -243,16 +243,38 @@ int net_device_table_mgr::map_net_devices()
...
+				// Close the cma_id which will not be offload
+				IF_RDMACM_FAILURE(rdma_destroy_id(cma_id)) {
+					ndtm_logerr("Failed in rdma_destroy_id (errno=%d %m)", errno);
+				} ENDIF_RDMACM_FAILURE;
+				continue;
```
Comment (OphirMunk) [inline src/vma/dev/net_device_table_mgr.cpp]: Missing a WARNING message here about not offloading this bond interface...

File: src/vma/dev/net_device_table_mgr.cpp (hunk 2 above)
Comment (OphirMunk) [inline src/vma/dev/net_device_table_mgr.cpp]: I would suggest something like this:
```
if (ifa->ifa_flags & IFF_MASTER) {
    valid = verify_bond_ipoib_or_eth_qp_creation(...);  // Inside this function may print WARNINGs
}
else {
    valid = verify_ipoib_or_eth_qp_creation(...);
}

if (!valid) {
    ...
}
```

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/317/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/335/ for details (Mellanox internal link).

**Comment by OphirMunk** [comment]
NO immediate action:
I will change commit message during commit to:
issue: 843662 on bond interface verify ipoib or eth qp creation on all slaves

**Comment by OphirMunk** [comment]
:+1:

**Comment by orkmellanox** [comment]
Any comment before I merge?

**Pull request merged**